cirq/contrib/quantum_volume/quantum_volume.py

```diff
@@ -91,45 +92,116 @@ def compute_heavy_set(circuit: cirq.Circuit) -> List[int]:
 
     """
+    mapping = compilation_result.swap_network.final_mapping()
+    circuit = compilation_result.swap_network.circuit
+
     # Add measure gates to the end of (a copy of) the circuit. Ensure that those
     # gates measure those in the given mapping, preserving this order.
     qubits = circuit.all_qubits()
     key = None
     if mapping:
+        # Add any qubits that were not explicitly mapped, so they aren't lost in
+        # the sorting.
+        for q in qubits:
+            if q not in mapping:
+                mapping[q] = q
         key = lambda q: mapping[q]
         qubits = frozenset(mapping.keys())
-    circuit_copy = circuit + cirq.measure(*sorted(qubits, key=key))
+
+    # Don't do a single large measurement gate because then the key will be one
+    # large string. Instead, do a bunch of single-qubit measurement gates so we
+    # preserve the qubit keys.
+    sorted_qubits = sorted(qubits, key=key)
+    circuit_copy = circuit + [cirq.measure(q) for q in sorted_qubits]
 
     # Run the sampler to compare each output against the Heavy Set.
-    measurements = sampler.run(program=circuit_copy, repetitions=repetitions)
+    trial_result = sampler.run(program=circuit_copy, repetitions=repetitions)
+
+    # Post-process the results, e.g. to handle error corrections.
+    results = process_results(mapping, compilation_result.parity_map,
+                              trial_result)
 
+    # Aggregate the results into bit-strings (since we are using individual
+    # measurement gates).
+
+    results = results.agg(lambda meas: cirq.value.big_endian_bits_to_int(meas),
+                          axis=1)
     # Compute the number of outputs that are in the heavy set.
-    num_in_heavy_set = np.sum(np.in1d(measurements.data.iloc[:, 0], heavy_set))
+    num_in_heavy_set = np.sum(np.in1d(results, heavy_set))
+
+    # Return the number of Heavy outputs over the number of valid runs.
+    return num_in_heavy_set / len(results)
+
 
-    # Return the number of Heavy outputs over the number of runs.
-    return num_in_heavy_set / repetitions
+def process_results(mapping: Dict[cirq.Qid, cirq.Qid],
+                    parity_mapping: Dict[cirq.Qid, cirq.Qid],
+                    trial_result: cirq.TrialResult) -> pd.DataFrame:
+    """Checks the given results for parity and throws away all of the runs that
+    don't pass the parity test.
+
+    Args:
+        mapping: The circuit's mapping from logical qubit to physical qubit.
+        parity_mapping: The mapping from result qubit to its parity qubit.
+        trial_result: The results to process.
+
+    Returns:
+        Returns the rows that passed the parity test, with the parity qubit
+        measurements removed.
+
+    """
+    bad_measurements = set()
+    # The circuit's mapping from physical qubit to logical qubit.
+    inverse_mapping: Dict[cirq.Qid, cirq.Qid] = {
+        v: k for k, v in mapping.items()
+    }
+
+    for final_qubit, original_qubit in mapping.items():
```

Comment:
Does this work?

```
data = trial_result.data
bad_measurements = set()
for q, pf in parity_mapping.items():
    p = inverse_mapping[pf]
    mismatches = np.nonzero(data[str(q)] != data[str(p)])
    bad_measurements.update(mismatches)
```

Or more involved:

```
data = trial_result.data
kept = np.ones(data.shape[0], dtype=np.bool)
for q, pf in parity_mapping.items():
    p = inverse_mapping[pf]
    kept &= data[str(q)] == data[str(p)]
return data.where(kept)
```

Reply:
Yes, something similar to that does. Thanks, much cleaner now!